src/Transformations.jl

```diff
@@ -44,4 +45,62 @@ function transform!(tn::TensorNetwork, ::HyperindConverter)
     end
 end
 
-# TODO column reduction, diagonal reduction, rank simplification, split simplification
+struct DiagonalReduction <: Transformation end
+
+function transform!(tn::TensorNetwork, ::DiagonalReduction; output_inds=nothing, atol=1e-12)
+    output_inds = output_inds === nothing ? openinds(tn) : output_inds
+    queue = collect(keys(tn.tensors))
+
+    while !isempty(queue) # loop over all tensors
+        idx = pop!(queue)
+        tensor = tn.tensors[idx]
+
+        diag_axes = find_diag_axes(parent(tensor), atol)
+
+        while !isempty(diag_axes) # loop over all diagonal axes
+            (i, j) = pop!(diag_axes)
+            ix_i, ix_j = labels(tensor)[i], labels(tensor)[j]
+
+            # do not reduce output indices
+            new, old = (ix_j in output_inds) ? ((ix_i in output_inds) ? continue : (ix_j, ix_i)) : (ix_i, ix_j)
+
+            # replace old index in the other tensors in the network
+            for other_idx in setdiff(keys(tn.tensors), idx)
+                other_tensor = tn.tensors[other_idx]
+                if old in labels(other_tensor)
+                    new_tensor = replace(other_tensor, old => new)
+                    tn.tensors[other_idx] = new_tensor
+                end
+            end
+
+            repeated_labels = replace(collect(labels(tensor)), old => new) |> join
+            removed_label = filter(l -> l != old, labels(tensor)) |> join
+
+            ein_code_str = "ein\"$repeated_labels->$removed_label\""
+            data = eval(Meta.parse("$(ein_code_str)($tensor)"))
```

Comment:
Don't call the compiler with `eval` and `Meta.parse`. This slows downs a lot. Use `EinCode` like I wrote in `contract`.